internal/daemon/controller/handlers/scopes/scope_service.go

```diff
@@ -355,6 +367,52 @@ func (s Service) DeleteScope(ctx context.Context, req *pbs.DeleteScopeRequest) (
 	return nil, nil
 }
 
+// ListKeys implements the interface pbs.ScopeServiceServer.
+func (s Service) ListKeys(ctx context.Context, req *pbs.ListKeysRequest) (*pbs.ListKeysResponse, error) {
+	if req.GetId() == "" {
+		req.Id = scope.Global.String()
+	}
+	if err := validateListKeysRequest(req); err != nil {
+		return nil, err
+	}
+	authResults := s.authResult(ctx, req.GetId(), action.ListScopeKeys)
+	if authResults.Error != nil {
+		return nil, authResults.Error
+	}
+
+	items, err := s.kmsRepo.ListKeys(ctx, req.GetId())
+	if err != nil {
+		if errors.Is(err, dbw.ErrRecordNotFound) {
+			return nil, handlers.NotFoundErrorf("unknown scope_id %q", req.Id)
+		}
+		return nil, err
+	}
+
+	finalItems := make([]*pb.Key, 0, len(items))
+	res := perms.Resource{
+		Type: resource.Scope,
+	}
+	for _, item := range items {
+		res.Id = item.Id
+		res.ScopeId = item.Scope
+
+		outputFields := authResults.FetchOutputFields(res, action.ListScopeKeys).SelfOrDefaults(authResults.UserId)
+		outputOpts := make([]handlers.Option, 0, 3)
+		outputOpts = append(outputOpts, handlers.WithOutputFields(&outputFields))
+		if outputFields.Has(globals.ScopeField) {
+			outputOpts = append(outputOpts, handlers.WithScope(authResults.Scope))
+		}
+
+		protoItem, err := keyToProto(ctx, item, outputOpts...)
+		if err != nil {
+			return nil, err
+		}
+		finalItems = append(finalItems, protoItem)
+	}
```

Comment:
Note that there is no filtering of keys based on authorization. You are either allowed to list all keys in a scope or no keys. The authorization is on the scope, not the keys.